R/particle_filter_state.R

```diff
@@ -31,28 +31,134 @@ particle_filter_state <- R6::R6Class(
     min_log_likelihood = NULL,
     support = NULL,
 
-    run_compiled = function() {
+    step_r = function(step_index, partial = FALSE) {
```

Comment:
Similar to the comment on the dust PR, what's the reason for separating deterministic and stochastic, rather than having a single step function?

Reply:
there's always a design trade-off here - one function with many if/elses or two functions with similar logic. Here the compiled and non-compiled versions are so different that the two functions approach feels much easier to follow the logic through. In particular the R step function is very complicated and has quite a few performance-critical tweaks